honeycomb-core/src/coords.rs

```diff
@@ -0,0 +1,254 @@
+//! Custom spatial representation
+//!
+//! This module contains all code used to model spatial representation and
+//! operations. We re-implement these basic structures in order to have
+//! better control over our structure.
+
+// ------ IMPORTS
+
+use std::ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Sub, SubAssign};
+
+// ------ CONTENT
+
+cfg_if::cfg_if! {
+    if #[cfg(feature = "single_precision")] {
+        pub type FloatType = f32;
+    } else {
+        pub type FloatType = f64;
+    }
+}
+
+/// 2-dimensional coordinates structure
+///
+/// The floating type used for coordinate representation is determined
+/// using feature and the [FloatType] alias.
+///
+/// # Example
+///
+/// ```rust
+/// use honeycomb_core::{Coords2, FloatType};
+///
+/// let unit_x = Coords2::unit_x();
+/// let unit_y = Coords2::unit_y();
+///
+/// assert_eq!(unit_x.dot(&unit_y), 0.0);
+/// assert_eq!(unit_x.normal_dir(), unit_y);
+///
+/// let two: FloatType = 2.0;
+/// let x_plus_y: Coords2 = unit_x + unit_y;
+///
+/// assert_eq!(x_plus_y.norm(), two.sqrt());
+/// assert_eq!(x_plus_y.unit_dir(), Coords2::from((1.0 / two.sqrt(), 1.0 / two.sqrt())));
+/// ```
+///
+#[derive(Debug, Clone, Copy, Default, PartialEq)]
+pub struct Coords2 {
+    /// First coordinate
+    pub x: FloatType,
+    /// Second coordinate
+    pub y: FloatType,
+}
```

Comment:
I would prefer a templated version of `Coords2` and not have a general switch and only one data type.